File: packages/pluto-py/pluto_client/secret.py
from pluto_base.resource import (
    IResource,
    IResourceCapturedProps,
    IResourceClientApi,
    IResourceInfraApi,
)
from pluto_base.platform import PlatformType
from pluto_base import utils


class ISecretClientApi(IResourceClientApi):
    def get(self) -> str:
        """Get the secret value."""
        raise NotImplementedError


class ISecretInfraApi(IResourceInfraApi):
    pass


class ISecretCapturedProps(IResourceCapturedProps):
    pass


class ISecretClient(ISecretClientApi, ISecretCapturedProps):
    pass


class ISecretInfra(ISecretInfraApi, ISecretCapturedProps):
    pass


class Secret(IResource, ISecretClient, ISecretInfra):
    fqn = "@plutolang/pluto.Secret"

    def __init__(self, name: str, value: str):
        platform_type = utils.current_platform_type()
        if platform_type == PlatformType.AWS:
            from .clients import aws

            self._client = aws.Secret(name, value)
        else:
            raise ValueError(f"not support this runtime '{platform_type}'")

File: packages/pluto/src/secret.ts
import {
  IResource,
  IResourceCapturedProps,
  IResourceClientApi,
  IResourceInfraApi,
  PlatformType,
  utils,
} from "@plutolang/base";
import { aws } from "./clients";

export interface ISecretClientApi extends IResourceClientApi {
  /**
   * Get the secret value.
   */
  get(): Promise<string>;
}

export interface ISecretInfraApi extends IResourceInfraApi {}
export interface ISecretCapturedProps extends IResourceCapturedProps {}

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * client implementation class of a resource type.
 */
export type ISecretClient = ISecretClientApi & ISecretCapturedProps;

/**
 * Construct a type that includes all the necessary methods required to be implemented within the
 * infrastructure implementation class of a resource type.
 */
export type ISecretInfra = ISecretInfraApi & ISecretCapturedProps;

export class Secret implements IResource {
  constructor(name: string, value: string) {
    name;
    value;
    throw new Error(
      "Cannot instantiate this class, instead of its subclass depending on the target runtime."
    );
  }

  public static buildClient(name: string, value: string): ISecretClient {
    const platformType = utils.currentPlatformType();
    switch (platformType) {
      case PlatformType.AWS:
        return new aws.Secret(name, value);
      default:
        throw new Error(`not support this runtime '${platformType}'`);
    }
  }

  public static fqn = "@plutolang/pluto.Secret";
}

export interface Secret extends IResource, ISecretClient, ISecretInfra {}
